src/ngx_http_lua_semaphore.c

```diff
@@ -0,0 +1,523 @@
+    {
+        /* load <= 50% and it's the older */
+        iter = (ngx_http_lua_semaphore_t*) (block+1);
+        for (i = 0; i < mm->num_per_block; i++, iter++) {
+            ngx_queue_remove(&iter->chain);
+        }
+
+        dd("free semaphore block: %p at procedure",block);
+
+        mm->total -= mm->num_per_block;
+        ngx_free(block);
+    }
+
+    dd("ngx_http_lua_free_semaphore sem: %p "
+       "sem->chain.prev: %p sem->chain.next: %p",
+       sem, sem->chain.prev, sem->chain.next);
+}
+
+
+static ngx_int_t
+ngx_http_lua_semaphore_resume(ngx_http_request_t *r)
+{
+    lua_State                   *vm;
+    ngx_connection_t            *c;
+    ngx_int_t                    rc;
+    ngx_http_lua_ctx_t          *ctx;
+
+    ctx = ngx_http_get_module_ctx(r, ngx_http_lua_module);
+    if (ctx == NULL) {
+        return NGX_ERROR;
+    }
+
+    ctx->resume_handler = ngx_http_lua_wev_handler;
+
+    c = r->connection;
+    vm = ngx_http_lua_get_lua_vm(r, ctx);
+
+    if (ctx->cur_co_ctx->sem_resume_status == SEMAPHORE_WAIT_SUCC) {
+         lua_pushboolean(ctx->cur_co_ctx->co, 1);
+         lua_pushnil(ctx->cur_co_ctx->co);
+
+    } else {
+         lua_pushboolean(ctx->cur_co_ctx->co, 0);
+         lua_pushstring(ctx->cur_co_ctx->co, "timeout");
+    }
+
+    rc = ngx_http_lua_run_thread(vm, r, ctx, 2);
+
+    ngx_log_debug1(NGX_LOG_DEBUG_HTTP, r->connection->log, 0,
+                   "lua run thread returned %d", rc);
+
+    if (rc == NGX_AGAIN) {
+        return ngx_http_lua_run_posted_threads(c, vm, r, ctx);
+    }
+
+    if (rc == NGX_DONE) {
+        ngx_http_lua_finalize_request(r, NGX_DONE);
+        return ngx_http_lua_run_posted_threads(c, vm, r, ctx);
+    }
+
+    /* rc == NGX_ERROR || rc >= NGX_OK */
+    if (ctx->entered_content_phase) {
+        ngx_http_lua_finalize_request(r, rc);
+        return NGX_DONE;
+    }
+
+    return rc;
+}
+
+
+int
+ngx_http_lua_ffi_semaphore_new(ngx_http_lua_semaphore_t **psem,
+    int n, char **errstr)
+{
+    ngx_http_lua_semaphore_t *sem;
+
+    sem = ngx_http_lua_alloc_semaphore();
+    if (sem == NULL) {
+        *errstr = "ngx_http_lua_ffi_semaphore_new ngx_alloc failed";
+        return NGX_ERROR;
+    }
+
+    ngx_queue_init(&sem->wait_queue);
+
+    sem->resource_count = n;
+    sem->wait_count = 0;
+    sem->in_post_event = 0;
+    sem->log = ngx_cycle->log;
+    *psem = sem;
+
+    dd("ngx_http_lua_ffi_semaphore_new semaphore: %p, resource_count: %d",
+       sem, sem->resource_count);
+
+    return NGX_OK;
+}
+
+
+int
+ngx_http_lua_ffi_semaphore_post(ngx_http_lua_semaphore_t *sem,
+    int n, char **errstr)
+{
+    if (sem == NULL) {
+        *errstr = "semaphore is null";
+        return NGX_ERROR;
+    }
+
+    sem->resource_count += n;
+
+    dd("ngx_http_lua_ffi_semaphore_post semaphore: %p, resource_count: %d",
+       sem, sem->resource_count);
+
+    if (!sem->in_post_event && !ngx_queue_empty(&sem->wait_queue)) {
```

Comment:
I'm afraid `in_post_event` is not a good name here. Grammatically, it should be "in_posted_event". But semantically this flag is to avoid posting duplicate events, I think `event_posted` is better. Thoughts?